In [1]:
from openai import OpenAI
from dotenv import load_dotenv
load_dotenv()

True

In [2]:
client = OpenAI()

In [3]:
# STEP 1: Creating the Assistant

assistant = client.beta.assistants.create(
    name="Mortgage Calculator",
    instructions="You use Python code to help answer questions about mortgage and interest payments.",
    tools = [{'type':'code_interpreter'}],
    model = 'gpt-3.5-turbo'
)

In [4]:
assistant.id

'asst_WqoaI7xLibpUW77eBUo0zdeL'

In [37]:
# STEP 2: Creating a Thread
thread = client.beta.threads.create()

In [38]:
thread.id

'thread_QPoNXF5dJZlQXHJANS7CdfmT'

In [35]:
delete_thread = client.beta.threads.delete('thread_86zxuRANGqlWXKKeRQ1RBYzJ')

In [36]:
delete_thread

ThreadDeleted(id='thread_86zxuRANGqlWXKKeRQ1RBYzJ', deleted=True, object='thread.deleted')

In [39]:
# STEP 3: Add message to thread
message_1 = client.beta.threads.messages.create(
    thread_id = thread.id,
    role = 'user',
    content = 'I want to buy a house that costs £1.5 Million on a 30 year fixed loan at 5% interest. What will my monthly payments be?')

In [47]:
# STEP 4: Run the thread w/ Assistant

import time

def run_and_wait(assistant_id, thread_id, instructions=""):
    run = client.beta.threads.runs.create(
        assistant_id = assistant_id,
        thread_id = thread_id,
        instructions = instructions
    )
    while run.status == 'queued' or run.status == 'in_progress':
        run = client.beta.threads.runs.retrieve(
            thread_id = thread.id,
            run_id = run.id)
        time.sleep(2)
        print(run.status)
    return run

In [48]:
run = run_and_wait(
    assistant_id = assistant.id,
    thread_id = thread.id,
    instructions = 'Give a detailed analysis and considerations regarding mortgage payments.'
)

in_progress
completed


In [43]:
# STEP 5: Display response
messages = client.beta.threads.messages.list(
    thread_id = thread.id,
    order = 'asc',
    after = message_1.id
)

In [44]:
def display_thread_messages(messages):
    for thread_message in messages.data:
        print(thread_message.content[0].text.value)
        print('\n')

In [45]:
display_thread_messages(messages)

To calculate the monthly mortgage payment for a loan, we can use the formula for calculating the monthly payment on a fixed-rate mortgage:

\[
M = P \times \frac{r(1+r)^n}{(1+r)^n - 1}
\]

Where:
- \( M \) is the monthly mortgage payment
- \( P \) is the principal loan amount (initial loan amount) - in this case, £1.5 Million
- \( r \) is the monthly interest rate (annual interest rate divided by 12)
- \( n \) is the number of monthly payments (loan term in years multiplied by 12)

Let's calculate the monthly mortgage payment based on the given loan amount, interest rate, and loan term.


The monthly mortgage payment for a £1.5 Million house on a 30-year fixed loan with a 5% interest rate would be approximately £8,052.32. 

This monthly payment includes both the principal amount and the interest. Over the 30-year term, you would make a total of 360 payments (12 payments per year for 30 years).

This calculation assumes that the interest rate remains fixed for the entire term of the loa

In [46]:
messages.data[0].content[0].text.value


"To calculate the monthly mortgage payment for a loan, we can use the formula for calculating the monthly payment on a fixed-rate mortgage:\n\n\\[\nM = P \\times \\frac{r(1+r)^n}{(1+r)^n - 1}\n\\]\n\nWhere:\n- \\( M \\) is the monthly mortgage payment\n- \\( P \\) is the principal loan amount (initial loan amount) - in this case, £1.5 Million\n- \\( r \\) is the monthly interest rate (annual interest rate divided by 12)\n- \\( n \\) is the number of monthly payments (loan term in years multiplied by 12)\n\nLet's calculate the monthly mortgage payment based on the given loan amount, interest rate, and loan term."

In [52]:
# Step 6: Create a new message for the same thread
message_2 = client.beta.threads.messages.create(
    thread_id = thread.id,
    role = 'user',
    content = 'What if I put a down payment of £700.000, how would this change my monthly payment?'
)

In [53]:
run = run_and_wait(
    assistant_id = assistant.id,
    thread_id = thread.id,
    instructions = 'Give a detailed analysis and considerations regarding mortgage payments.'
)

in_progress
in_progress
in_progress
in_progress
in_progress
in_progress
in_progress
in_progress
in_progress
in_progress
in_progress
in_progress
completed


In [56]:
messages = client.beta.threads.messages.list(
    thread_id = thread.id,
    order = 'asc',
    after = message_1.id
)

In [57]:
display_thread_messages(messages)

To calculate the monthly mortgage payment for a loan, we can use the formula for calculating the monthly payment on a fixed-rate mortgage:

\[
M = P \times \frac{r(1+r)^n}{(1+r)^n - 1}
\]

Where:
- \( M \) is the monthly mortgage payment
- \( P \) is the principal loan amount (initial loan amount) - in this case, £1.5 Million
- \( r \) is the monthly interest rate (annual interest rate divided by 12)
- \( n \) is the number of monthly payments (loan term in years multiplied by 12)

Let's calculate the monthly mortgage payment based on the given loan amount, interest rate, and loan term.


The monthly mortgage payment for a £1.5 Million house on a 30-year fixed loan with a 5% interest rate would be approximately £8,052.32. 

This monthly payment includes both the principal amount and the interest. Over the 30-year term, you would make a total of 360 payments (12 payments per year for 30 years).

This calculation assumes that the interest rate remains fixed for the entire term of the loa

To calculate the monthly mortgage payment for a loan, we can use the formula for calculating the monthly payment on a fixed-rate mortgage:

$$
M = P \times \frac{r(1+r)^n}{(1+r)^n - 1}
$$

Where:
- \( M \) is the monthly mortgage payment
- \( P \) is the principal loan amount (initial loan amount) - in this case, £1.5 Million
- \( r \) is the monthly interest rate (annual interest rate divided by 12)
- \( n \) is the number of monthly payments (loan term in years multiplied by 12)

Let's calculate the monthly mortgage payment based on the given loan amount, interest rate, and loan term.


The monthly mortgage payment for a £1.5 Million house on a 30-year fixed loan with a 5% interest rate would be approximately £8,052.32. 

This monthly payment includes both the principal amount and the interest. Over the 30-year term, you would make a total of 360 payments (12 payments per year for 30 years).

This calculation assumes that the interest rate remains fixed for the entire term of the loan and does not take into account additional costs such as property taxes, insurance, or other fees that may be associated with your mortgage.


Let me know if you need any further assistance!


What if I put a down payment of £700.000, how would this change my monthly payment?


If you make a down payment of £700,000 towards the purchase price of £1.5 Million, it will reduce the principal amount of the loan. The new principal amount would be £1.5 Million - £700,000 = £800,000.

We can now recalculate the monthly mortgage payment using the reduced principal amount. Let's calculate the new monthly payment.


With a down payment of £700,000, the new monthly mortgage payment for a £1.5 Million house on a 30-year fixed loan with a 5% interest rate would be approximately £4,294.57. 

By making a larger down payment, you effectively reduce the principal amount of the loan, resulting in a lower monthly payment. The down payment reduces the amount you need to borrow, which leads to lower monthly interest payments and a shorter loan term.

Feel free to reach out if you have any more questions or need further assistance!


In [58]:
run_steps = client.beta.threads.runs.steps.list(
    thread_id = thread.id,
    run_id = run.id
)

In [59]:
for step in run_steps:
    print(step.step_details)
    print('\n')

MessageCreationStepDetails(message_creation=MessageCreation(message_id='msg_rqHIhFMTE9RmJ72pFDP4oKCw'), type='message_creation')


ToolCallsStepDetails(tool_calls=[CodeInterpreterToolCall(id='call_7tT4czKCPWamADfJ9Mpnipor', code_interpreter=CodeInterpreter(input='# Given values\nprincipal = 1500000  # loan amount in pounds\nannual_interest_rate = 0.05  # 5% annual interest rate\nloan_term_years = 30  # 30 year loan term\ndown_payment = 700000  # down payment in pounds\n\n# New principal amount after down payment\nnew_principal = principal - down_payment\n\n# Convert annual interest rate to monthly interest rate\nmonthly_interest_rate = annual_interest_rate / 12\n\n# Convert loan term to months\nloan_term_months = loan_term_years * 12\n\n# Calculate new monthly mortgage payment using the formula\nnew_monthly_payment = new_principal * (monthly_interest_rate * (1 + monthly_interest_rate)**loan_term_months) / ((1 + monthly_interest_rate)**loan_term_months - 1)\nnew_monthly_payment', output

In [21]:
# Delete the assistant:
response = client.beta.assistants.delete(assistant.id)

In [22]:
response

AssistantDeleted(id='asst_HYSt1vSEPMsIFUrvgvn4bqVy', deleted=True, object='assistant.deleted')